rfcs/020 - Dependencies permissions RFC.md

```diff
@@ -0,0 +1,168 @@
+
+- Define which dependencies or standard libraries will be disabled for all dependencies. Meaning if I'm installing 2 external dependencies, and both of them use `@std/cmd` and I want to block it, I can `disable` `@std/cmd` and will be block for both of them. It will be translated in mocks at `./depeendencies/modules/` level.
+
+![Block at library level](./lib-lvl-v1.png)
+
+![Block at app level](./app-lvl-v1.png)
+
+- Define which external libraries will be part of the project and we could add specific blocks for dependecies used in the external library.
+
+The `@std/deps` will have the following API:
+
+```ts
+type Dependency {
+  url: string, // Empty string or ignore it for global blocks at application level
+  block: Array<string>,
+  fullBlock: Array<string>,  // fullBlock, deepBlock, extensiveBlock, exhaustiveBlock?
+}
+
+type StdDependency { 
+  name: string,
+}
+
+fn use(std: string) = new StdDependency {
+  name: std,
+};
+
+fn add(url: string) = new Dependency {
+  url: url,
+  block: [],
+  fullBlock: [],
+};
+
+fn block(block: string): Dependency {
+  return new Dependency {
+    url: '',
+    block: [block],
+    fullBlock: [],
+  };
+}
+
+fn block(block: Array<string>): Dependency {
+  return new Dependency {
+    url: '',
+    block: block,
+    fullBlock: [],
+  };
+}
+
+fn block(dep: Dependency, block: string): Dependency {
+  dep.block.push(block);
+  return dep;
+}
+
+fn block(dep: Dependency, block: Array<string>): Dependency {
+  dep.block = dep.block + block;
+  return dep;
+}
+
+fn fullBlock(dep: Dependency, block: string): Dependency {
+  dep.fullBlock.push(fullBlock);
+  return dep;
+}
+
+fn fullBlock(dep: Dependency, block: Array<string>): Dependency {
+  dep.fullBlock = dep.fullBlock + block;
+  return dep;
+}
+
+fn commit(using: Array<StdDependency>, dependencies: Array<Dependency>) {
+  // Remove std dep from blacklist
+  // Download and install each dep
+  // Apply blocks defined for each dependency
+}
+fn commit(using: Array<StdDependency>, dependencies: Array<Dependency>, global: Array<Dependency>) {
+  // Remove std dep from blacklist
+  // Download and install each dep
+  // Apply blocks defined for each dependency
+  // Block/Enable (based on what we finally decide) depenencies at application level, meaning that mocks will exists at /dependencies/modules/
+}
+```
+The `.dependencies.ln` file could look something like:
+
+```ts
+const using = [
+  use('@std/tcp');
+];
+const dependencies = [
+  // Block `@std/cmd` but do not override if any mock exists:
+  add('https://github.com/org/new_dep_A').block('@std/cmd'),
+  // Block `@std/cmd` and override mocks if any:
+  add('https://github.com/org/new_dep_B').fullBlock('@std/cmd'),
+  // Do not block any standrad library:
+  add('https://github.com/org/new_dep_C'),
+];
+const globalDepsBlock = [
+  block('@std/fs'),
+  fullBlock('@std/exec'),
+];
+commit(using, dependencies, globalDepsBlock);
```

Comment:
Another idea here:

```js
using([
  '@std/tcp',
]).dependencies([
  add('https://github.com/org/new_dep_A').block('@std/cmd'),
  add('https://github.com/org/new_dep_B').fullBlock('@std/cmd'),
  add('https://github.com/org/new_dep_C'),
]).globalDepsBlock([
  block('@std/fs'),
  fullBlock('@std/exec'),
]).commit();
```

Or maybe:

```js
module()
  .using('@std/tcp')
  .dependency('https://github.com/org/new_dep_A')
    .block('@std/cmd')
    .add()
  .dependency('https://github.com/org/new_dep_B')
    .fullBlock('@std/cmd')
    .add()
  .dependency('https://github.com/org/new_dep_C')
    .add()
  .globalBlock('@std/fs')
  .globalFullBlock('@std/exec')
  .commit();
```

Reply:
I don't know about the `global` naming, either. You could potentially stick to `block` and `fullBlock` and it applying to this library/module versus a dependency depends on whether it's attached to a dependency declaration or the library/module declaration.

Reply:
Agree, add that name to distinguish function parameters but using an API ad the proposed one makes sense to just apply to the module or dependency the block/fullBlock functions.